### PR TITLE
Game browser

### PR DESCRIPTION
**Pull request by BenNF**
Added game browser component visible at the /games route as per #91.

The browser loads the 21 most popular games and all their data and displays them in tiles similar to the Game component.

The component was put in the landing module as it was already dealing with popular games. 

For the this component the API was extended on the /popular endpoint to allow for a limit parameter.
The API endpoint /games/mulit endpoint was added the accepts a list of slugs returns game data for all of them.

The game browser is not currently but could be displayed in the App component once a user is authenticated. 

Feedback is totally welcome let me know what you think.

image

### CODE REVIEW

**Comment by BenNF** [comment]
Not sure why this is still failing on tests. I added a test for the new endpoint and the old one is already covered. Any ideas?

Additionally looks like there is some duplicate code with #100 in the Details component. Both PR's add handling for when data is undefined so we'll have to pick one of them.

**Comment by danielgrijalva** [comment]
@BenNF apparently it's failing because the coverage would decrease if merged into master. The diff is only 0.06% so maybe you're not testing a line or two? 

Running `coverage run backend/manage.py test games actions users` from the project root and then `coverage html` will generate an interactive report where you can check the specific lines that are being tested.

**Comment by danielgrijalva** [comment]
By the way, this is a great addition! it's certainly one of the biggest features of Overworld, so it will require a lot of work to get it _right_. First thing that comes to mind is adding pagination or a "load more" button at the bottom, as well as filters for browsing by other criteria like platforms, genres, etc.

For the time being, we have to work on the design. I would implement the tiles with the [Card](https://react.semantic-ui.com/views/card/) component from Semantic UI, using the cover as the image.

I feel that using the `<Details />` component is a bit too much, makes the feed look too cluttered. The same thing with the game's description (perhaps it can be shortened more with a "read more" link that goes to the game's page?).

Also, in each game card I would include some quick informative info (I'm not sure _what_ yet). I'm looking at how [rawg.io](https://rawg.io/) displays games for some inspiration...

**Comment by BenNF** [comment]
@danielgrijalva Thanks for all the feedback. I'll take a first pass at making some changes based on it tomorrow including load more, filters, and changing it to use more of the semantic ui components.

In the meantime I just added two tests which I think get the edge cases that codecov was complaining about.

**Comment by BenNF** [comment]
Updated with a first pass at filtering as well as some redesign using card components and pagination. 

The card components currently only show title, cover, developer, release year, and a shortened summary.

Next steps are extending api to search for developers, genres + any other fields we would like to filter by,  as well as extending the /games/popular endpoint to take filter parameters.

With Filters: 
image


Without Filters: 
image


Feedback is appreciated as always :)

Sidenote: IGDB seems to serve a lot of NSFW games(as you can see in the screenshots) when sorting by popularity(as the /games/popular endpoint does) and there doesn't seem to be an easy way to filter them out by genre. @danielgrijalva Any ideas what to do about this? Do you want to leave them or try to find some way to filter them out?

File: frontend/src/modules/landing/components/gamebrowser/index.js [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+    if (game) {
+      const description = game.summary
+        ? game.summary.slice(0, 100) + "..."
+        : "";
+
+      const cover = game.cover
+        ? `https://images.igdb.com/igdb/image/upload/t_cover_big/${
+            game.cover.image_id
+          }.jpg`
+        : null; //TODO there should be a default cover if the API doesnt provide one
```
Comment (danielgrijalva) [inline frontend/src/modules/landing/components/gamebrowser/index.js]: Yeah I've thought of this... the "default" cover should simply be a placeholder with the game name, like this:
> image

But don't worry about it, I'm gonna open a new issue/PR for this...

File: frontend/src/modules/landing/components/gamebrowser/index.js [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    };
+    if (game) {
+      const description = game.summary
+        ? game.summary.slice(0, 100) + "..."
+        : "";
+
+      const cover = game.cover
+        ? `https://images.igdb.com/igdb/image/upload/t_cover_big/${
+            game.cover.image_id
+          }.jpg`
```
Comment (danielgrijalva) [inline frontend/src/modules/landing/components/gamebrowser/index.js]: Is it possible to use the `<Cover />` component instead of the raw image? That way, all the game cards could be the same size

File: frontend/src/modules/landing/components/gamebrowser/index.js [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+      return (
+        <React.Fragment>
+          {platforms.map(platform => {
+            let icon = null;
+
+            if (platform.name.includes("PC")) icon = faSteam;
+            else if (platform.name.includes("Playstation"))
+              icon = faPlaystation;
+            else if (platform.name.includes("Xbox")) icon = faXbox;
+            return <FontAwesomeIcon icon={icon} size={"sm"} />;
```
Comment (danielgrijalva) [inline frontend/src/modules/landing/components/gamebrowser/index.js]: You could also use the `<Icon />` component from Semantic UI to avoid the fontawesome imports!

**Comment by danielgrijalva** [comment]
I just provided some quick feedback, will be adding more comments throughout the day.

Regarding the NSFW games, I would prefer to not show them/filter them out. Maybe the `age_ratings` field can be of help here? [List of game fields](https://api-docs.igdb.com/?javascript#game)

By the way, a great addition the to games browser would be something like this:
> image

I believe that with the "most anticipated" section we can use the `hypes` field:
> `hypes`: number of follows a game gets before release

And the `status` or `release_dates` fields for the coming soon and recently released section. I'd be happy to help in this regard, just let me know!

File: backend/games/urls.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 urlpatterns = [
     path('popular/', views.get_popular_games, name='get-popular'),
+    path('multi/', views.get_games, name='get-games'),
```
Comment (danielgrijalva) [inline backend/games/urls.py]: Not sure if I'm a fan of this... an alternative would be to simply use `/api/games` for retrieving multiple games... e.g. `/api/games/?slugs=...`. This can be accomplished by moving the URL to the bottom of the urlpatterns and leave the pattern string empty :thinking:

File: frontend/src/modules/landing/components/gamebrowser/index.js [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+const mapStateToProps = state => ({
+  popular: state.landing.popular,
+  loading: state.landing.isLoadingPopular,
+  games: state.landing.gameData
+});
+
+export default connect(
+  mapStateToProps,
+  { getPopular, getGameData }
+)(GameBrowser);
```
Comment (danielgrijalva) [inline frontend/src/modules/landing/components/gamebrowser/index.js]: I'm thinking this component should be its own module... it's too big/complex to have it in one file inside the Landing module. So maybe all of this should be inside a `games/` or `game-browser/` folder? Obviously with the folder structure and files that the other modules have.

File: frontend/src/modules/landing/components/filters/genres.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
+
+//TODO these should probably be search from the server rather than hardcoded
```
Comment (danielgrijalva) [inline frontend/src/modules/landing/components/filters/genres.js]: Definitely! Calling `https://api-v3.igdb.com/genres` with a limit of 50 (max. for free API keys) returns all of them :+1:

File: frontend/src/modules/landing/components/gamebrowser/index.js (hunk 2 above)
Comment (BenNF) [reply frontend/src/modules/landing/components/gamebrowser/index.js]: The Semantic ui Card component will only take a url for the Image, I could switch to using the <Cover> component but we would have to stop using <Card> and remake the look ourselves. Do you think this is worth it to use <Cover>?

**Comment by BenNF** [comment]
@danielgrijalva just updated this based on some of your feedback.

In regards to NSFW games, age_ratings is not consistently populated in IGDB and is extremely inconstant on NSFW games. It doesn't look like that will be a way to filter them out.

I also have a few questions about the new game browser design you posted. Would that be replacing the existing view with <Card> components, or added as another way to browse games? 
How, if at all, is filtering involved in the new view?

Sorry I haven't been able to work on this that much this week.

**Comment by danielgrijalva** [comment]
Apparently, most NSFW games (at least the ones I see in the screenshot) have `Erotic` in their `themes` field. That can be a way to filter them, we definitely don't want erotic games on the front page :sweat_smile: 

Oh and it's definitely not a new design; it's an addition to the section you've been working on. It's definitely a new component, but it will be placed at the bottom of the game cards. By the way, it's not really necessary for this PR, I can work on it once we merge this one! (unless you've already started?). And I don't think this new section requires filters or anything like that. 

One more thing: due to this new section I think the "load more" button for loading more game cards is not necessary anymore? at least for the game browser front page. Maybe in the future we can have a dedicated page for showing popular games e.g. `/popular`, and have that "infinite scrolling" experience there.

**Comment by BenNF** [comment]
@danielgrijalva I've updated this with changes for most of your feedback points including some refactoring. 

There is more work to be done here adding the footer component, more filters, and I'm sure bugs will appear, but those seem like things for future PR's as this one is getting quite big. 

I'm feeling pretty good about where this is right now, and would love some feedback about what other features/fixes should be done in this PR.

**Comment by danielgrijalva** [comment]
I'm gonna merge this now. I'll be working on tweaking and fixing small visual details like colours, the CSS specificity and try to make all the game tiles the same size.
